# Determine gender in population data sets of *Salix nigra*
Nan Hu

Dec. 06, 2020

---

This idea was come up from the demand that we want to know the gender of Minghao's *Salix nigra* sample for further analysis.
However, because of late sampling season, it was unable to check individuals gender by that time. 
Therefore, here we designed this workflow to identify sex in this sampling population.

Here are packages used in this analysis:
* [GATK: Genome Analysis Toolkit]https://gatk.broadinstitute.org/hc/en-us
* 
**Step 1: Data Preparation**
This work starts from a filtered ```.vcf``` file generated by GATK software packages variants calling workflow. 
We will take non-missing positions from ```.vcf``` file as the input for BEAGLE 5.1 software.

Follwing scripts are removing unfiltered SNPs from the ```.vcf``` file to reduce the size of ```.vcf``` file.

```bash
#!/bin/sh
#$ -V
#$ -cwd
#$ -S /bin/bash
#$ -N exclude_filtered
#$ -q omni
#$ -o log/p00/$JOB_NAME.o$JOB_ID
#$ -e log/p00/$JOB_NAME.e$JOB_ID
#$ -pe sm 1
#$ -P quanah

source ~/.bashrc
gatk SelectVariants \
    -R ../../SE_SDR_v5_w/raw/ref/Salix_purpurea_var_94006.mainGenome.fasta \
    -V ../raw/vcf/snigra_minghao.Chr07.SLR.recode.vcf \
    --exclude-filtered TRUE \
    --max-nocall-fraction 0.25 \
    --sample-name allnames.txt \
    --allow-nonoverlapping-command-line-samples \
    -O ../raw/vcf/snigra_minghao.Chr07.SLR.recode.reduceNoCall0.25.vcf
 ```


**Step 2: Convert to BEAGLE format**



**Step 3: Gender Assignment and Output**


