# Determine gender in population data sets of *Salix nigra*
Nan Hu

Dec. 06, 2020

---

This idea was come up from the demand that we want to know the gender of Minghao's *Salix nigra* sample for further analysis.
However, because of late sampling season, it was unable to check individuals gender by that time. 
Therefore, here we designed this workflow to identify sex in this sampling population using SNPs called from target sequencing data.

Here are software used in this analysis:
* [GATK: Genome Analysis Toolkit](https://gatk.broadinstitute.org/hc/en-us)
* [BEAGLE 5.1](http://faculty.washington.edu/browning/beagle/beagle.html)
* [vcftools](http://vcftools.sourceforge.net/)

R packages:
* dplyr
* tidyr

## Step 1: Data Preparation
This work starts from a filtered ```.vcf``` file generated by GATK software packages variants calling workflow. 
We will take non-missing positions from ```.vcf``` file as the input for BEAGLE 5.1 software.

Since we only interest in sex determination region (sex linked region) in Salix nigra, data should be subset to specific region for downstream analysis.
```bash
#!/bin/sh
#$ -V
#$ -cwd
#$ -S /bin/bash
#$ -N extract_SLR
#$ -q omni
#$ -o log/p01/$JOB_NAME.o$JOB_ID
#$ -e log/p01/$JOB_NAME.e$JOB_ID
#$ -pe sm 18
#$ -P quanah

vcftools --vcf ./vcf/snigra.all.snps.vcf \
         --chr Chr07 \
         --from-bp 4880000 \
         --to-bp 6088000 \
         --recode \
         --out ./vcf/snigra.SLR
 ```
Output file will be ```snigra.SLR.recode.vcf```.

Follwing scripts are removing filtered SNPs from the ```.vcf``` file to reduce the size of ```.vcf``` file. vcftools could do the same job but in order to be consistent from filtering steps, we suggest to use ```gatk SelectVariants``` to remove the redundant rows.
If reference genome has not been indexed and dictionaried. Please run [these commands](https://gatk.broadinstitute.org/hc/en-us/articles/360035531652-FASTA-Reference-genome-format) to prepare.
```allnames.txt``` is a list of individual names in data set. We add ```--allow-nonoverlapping-command-line-samples``` option here so the list can be different from individuals names in ```.vcf``` file. It will only bring up names overlapped.
```bash
#!/bin/sh
#$ -V
#$ -cwd
#$ -S /bin/bash
#$ -N exclude_filtered
#$ -q omni
#$ -o log/p02/$JOB_NAME.o$JOB_ID
#$ -e log/p02/$JOB_NAME.e$JOB_ID
#$ -pe sm 1
#$ -P quanah

gatk SelectVariants \
    -R ./ref/Salix_purpurea_var_94006.mainGenome.fasta \
    -V ./vcf/snigra.SLR.recode.vcf \
    --exclude-filtered TRUE \
    --max-nocall-fraction 0.25 \
    --sample-name allnames.txt \
    --allow-nonoverlapping-command-line-samples \
    -O ./vcf/snigra.SLR.reduced.vcf
```
The output of these code will be ```snigra.SLR.reduced.vcf```.

Since our aim is to find a set of SNPs consistantly having the same genotypes in one gender and having distinct genotypes in another, it is fine for us to do some stringent filtering criterions. Here, we want to remove all the SNPs that containing missing genotype callings. (Although we could keep them, we still need to control the maxium nocall fractions. In previous step we set it to 0.25, which is safe enough and will not affect the gender determination work actually. However, here we only keep SNPs that every individuals having calls to minimize our input datasets.)
```bash
#!/bin/sh
#$ -V
#$ -cwd
#$ -S /bin/bash
#$ -N exclude_filtered
#$ -q omni
#$ -o log/p03/$JOB_NAME.o$JOB_ID
#$ -e log/p03/$JOB_NAME.e$JOB_ID
#$ -pe sm 1
#$ -P quanah

vcftools --vcf ./vcf/snigra.SLR.reduced.vcf \
         --max-missing 1.0 \
         --recode \
         --out ./vcf/snigra.SLR.reduced.nomissing
```
The output file will be ```snigra.SLR.reduced.nomissing.recode.vcf```. This is ready for BEAGLE software as input.

## Step 2: Convert to BEAGLE format
This step is using BEAGLE software to convert ```.vcf``` file into BEAGLE format. The BEAGLE format is a tablized genotype table by extracting ```gt``` field of ```.vcf``` file. Parameter ```ne``` is the effective population size. We use default settings here since we only want convert format, we use ```gt``` field as input to avoid unnecessary phasing process.
```bash
#!/bin/sh
#$ -V
#$ -cwd
#$ -S /bin/bash
#$ -N make_beagle_shape
#$ -q omni
#$ -o log/p04/$JOB_NAME.o$JOB_ID
#$ -e log/p04/$JOB_NAME.e$JOB_ID
#$ -pe sm 1
#$ -P quanah

beagle gt=./vcf/snigra.SLR.reduced.nomissing.recode.vcf \
       ne=100000 \
       nthreads=18 \
       out=./vcf/snigra.SLR.out
```
The output of this is ```snigra.SLR.out.vcf.gz```. We can use ```gunzip snigra.SLR.out.vcf.gz``` to unzip it.

Before analyzing the data, we have several cleaning steps to form a ```.csv``` file.
```bash
tail -n +10 snigra.SLR.out.vcf > snigra.SLR.out.vcf.noinfo
head -n 1 snigra.SLR.out.vcf.noinfo | sed 's/#CHROM/CHROM/g' > snigra_raw.header
tail -n +2 snigra.SLR.out.vcf.noinfo  | sed 's/\t/\",\"/g' | tr '\n' '"' | sed 's/Chr/\n\"Chr/g' > snigra_raw.body
cat snigra_raw.header snigra_raw.body | sed 's/\t/,/g' > snigra_raw.csv
```
This is the input file for Step 3 R scripts.

## Step 3: Gender Assignment and Output
In this step, we will take input table to run some statistical test to find the potential sex linked SNPs.
```R
library(dplyr)
library(tidyr)

# working directory
# setwd()

raw.snp <- read.csv("snigra_raw.csv", stringsAsFactors = FALSE)
head(raw.snp)

# remove redundant columns
# remove non-bialletic SNPs
raw.snp.drop <- raw.snp %>% 
  select(c(-CHROM, -ID, -REF, -QUAL, -FILTER, -INFO, -FORMAT)) %>%
  filter(!grepl(",", ALT))

# reshape the dataframe for counting genotypes
raw.snp.longer <- raw.snp.drop %>% 
  pivot_longer(cols = c(-POS, -ALT), names_to = "IDV")
head(raw.snp.longer)
```

We made an assumptions here. We assume the sex ration of *Salix nigra* is 1:1. Thus, we would expected if one SNP is related to sex, in our population, it should appear to have a ratio between homozygotic and heterozygotic genotype near 1:1. Thus, we could make a simple chi-square test on the ratio of homo:hetero = 1:1. We will filter out all the site that are significantly far away from this ratio (p < 0.05). The rest SNPs will be candidates for assigning gender.
```R
# count genotypes of each position
snp.geno.sum <- full_join(raw.snp.longer %>% 
                            filter(value == "0|0") %>% 
                            group_by(POS) %>% 
                            summarise(ref_homo = n()),
                      raw.snp.longer %>% 
                        filter(value == "1|1") %>% 
                        group_by(POS) %>% 
                        summarise(alt_homo = n()), by = "POS") %>%
  full_join(raw.snp.longer %>% 
              filter(value == "0|1" | value == "1|0") %>% 
              group_by(POS) %>% 
              summarise(het = n()), by = "POS")

snp.geno.sum[is.na(snp.geno.sum)] <- 0

# calculate expected genotype ratio
snp.geno.exp <- raw.snp.drop %>% 
  mutate(exp_homo = (length(raw.snp.drop[1, ]) - 2) / 2) %>% 
  mutate(exp_het = (length(raw.snp.drop[1, ]) - 2) / 2)

# do a chi-square test
snp.chisq <- left_join(snp.geno.exp, snp.geno.sum, by = "POS") %>% 
  mutate(act_homo = ref_homo + alt_homo) %>%
  rowwise() %>% 
  mutate(chisq_stat = chisq.test(c(act_homo, het))$statistic, 
         p_val = chisq.test(c(act_homo, het))$p.value)

# filter SNPs that p <= 0.05
# only extract SNPs that homozygotic sites only refer to "0|0" or "1|1", no either 
snp.p_gt_0.05 <- snp.chisq %>% 
  filter(p_val > 0.05) %>% 
  filter(ref_homo == 0 | alt_homo == 0)

# further cleaning, remove calculations
snp.pre_sex <- snp.p_gt_0.05 %>% 
  select(-ref_homo, -alt_homo, -exp_het, 
         -exp_homo, -act_homo, -chisq_stat, 
         -p_val, -het, -ALT)
```

We assume males are heterozygotes while females are all homozygotes on these SNPs. Then, we can check how many proportion of these SNPs agrees with males or females.
```R
# assign putative sex
snp.pre_sex[snp.pre_sex == "0|0"] = "F"
snp.pre_sex[snp.pre_sex == "1|1"] = "F"
snp.pre_sex[snp.pre_sex == "0|1"] = "M"
snp.pre_sex[snp.pre_sex == "1|0"] = "M"
head(snp.pre_sex)

# count "F"s and "M"s for each individuals
snp.pre_sex.longer <- snp.pre_sex %>% 
  pivot_longer(cols = c(-POS), names_to = "IDV")
agree_prop <- full_join(snp.pre_sex.longer %>% 
                          filter(value == "F") %>%
                          group_by(IDV) %>% 
                          summarise(pr.f = n()), 
                        snp.pre_sex.longer %>% 
                          filter(value == "M") %>% 
                          group_by(IDV) %>% 
                          summarise(pr.m = n()), by = "IDV")
agree_prop[is.na(agree_prop)] <- 0

# assign gender only when 90% of SNP agree with specific gender
sex.out.table <- agree_prop %>% 
  mutate(pr.female = pr.f / (pr.f + pr.m)) %>% 
  mutate(gender = ifelse(pr.female > 0.9, "F", 
                         ifelse(pr.female < 0.1, "M", "UnK"))) %>%
  select(IDV, gender)

# output file
write.csv(sex.out.table, "snigra.individual.sex.csv")
```
Now we will get a ```snigra.individual.sex.csv``` file contains gender information of each individuals. These results were examined by around 20 individuals known for sex.
